# Intro to Data Science Final Exam Supervised Learning Review

## Name: Belinda Gan

## NIM: 2440094680

In [1]:
print("\n")

# Setting Jupyter Notebook Themes

In [2]:
!pip install jupyterthemes

You should consider upgrading via the 'c:\users\belinda gan\anaconda3\python.exe -m pip install --upgrade pip' command.


In [3]:
!jt -l

Available Themes: 
   chesterish
   grade3
   gruvboxd
   gruvboxl
   monokai
   oceans16
   onedork
   solarizedd
   solarizedl


In [4]:
!jt -monokai

In [5]:
from jupyterthemes import get_themes
from jupyterthemes.stylefx import set_nb_theme

In [6]:
set_nb_theme('monokai')

In [7]:
print("\n\n")

# Data Dictionary

In [8]:
"""
For more information, read [Cortez et al., 2009].

Input variables (based on physicochemical tests):
1 - fixed acidity
2 - volatile acidity
3 - citric acid
4 - residual sugar
5 - chlorides
6 - free sulfur dioxide
7 - total sulfur dioxide
8 - density
9 - pH
10 - sulphates
11 - alcohol

Output variable (based on sensory data):
12 - quality (score between 0 and 10)
"""

'\nFor more information, read [Cortez et al., 2009].\n\nInput variables (based on physicochemical tests):\n1 - fixed acidity\n2 - volatile acidity\n3 - citric acid\n4 - residual sugar\n5 - chlorides\n6 - free sulfur dioxide\n7 - total sulfur dioxide\n8 - density\n9 - pH\n10 - sulphates\n11 - alcohol\n\nOutput variable (based on sensory data):\n12 - quality (score between 0 and 10)\n'

In [9]:
print("\n\n")

## Tips

In [10]:
"""
What might be an interesting thing to do, is aside from using regression modelling, 
is to set an arbitrary cutoff for your dependent variable (wine quality) at e.g. 7
or higher getting classified as 'good/1' and the remainder as 'not good/0'.
This allows you to practice with hyper parameter tuning on e.g. decision tree algorithms
looking at the ROC curve and the AUC value.
Without doing any kind of feature engineering or overfitting, 
you should be able to get an AUC of .88 (without even using random forest algorithm)
"""

"\nWhat might be an interesting thing to do, is aside from using regression modelling, \nis to set an arbitrary cutoff for your dependent variable (wine quality) at e.g. 7\nor higher getting classified as 'good/1' and the remainder as 'not good/0'.\nThis allows you to practice with hyper parameter tuning on e.g. decision tree algorithms\nlooking at the ROC curve and the AUC value.\nWithout doing any kind of feature engineering or overfitting, \nyou should be able to get an AUC of .88 (without even using random forest algorithm)\n"

In [11]:
print("\n\n")

# Import Libraries

In [12]:
import numpy as np # for statistical calculation and is basic of scipy library
import statistics as statistics # for statistical calculation
import pandas as pd # for Dataframe and scikit-learn library since scikit-learn will at least need numpy array or pandas dataframe
import matplotlib.pyplot as plt # for data visualization
import seaborn as sb # for data visualization
%matplotlib notebook
# to make matplotlib or seaborn graphs included in Jupyter notebook, directly below the code cell that generates it and add interactivity as well

# I use scikit-learn (sklearn) library to split data into training and testing dataset; predict the wine quality using various Supervised Learning algorithms (Logistic Regression, K-Nearest Neighbours, Naive Bayes, Decision Tree, Neural Network, Random Forest, Support Vector Machine; generate confusion matrix and classification report; standardize and normalize data in order to make standar deviation value closer to 0, which will make prediction more accurate
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler, normalize # to standardize and normalize data to make the dataset's standard deviation closer to 0 so that the prediction will be more accurate
from sklearn.linear_model import LogisticRegression # for Logistic Regression algorithm
from sklearn.neighbors import KNeighborsClassifier # for K-Nearest Neighbors algorithm
from sklearn.naive_bayes import GaussianNB # for Naive Bayes algorithm
from sklearn.neural_network import MLPClassifier # for Neural Network algorithm
from sklearn.tree import DecisionTreeClassifier # for Decision Tree algorithm
from sklearn.svm import SVC # for Support Vector Machine algorithm
from sklearn.ensemble import RandomForestClassifier # for Random Forest algorithm

In [13]:
print("\n\n")

# Data Overview

In [14]:
wineDataset = pd.read_csv("winequality-red.csv")
wineDataset

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6  

[1599 rows x 12 columns]

In [15]:
wineDataset.head(50)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0             7.4             0.700         0.00             1.9      0.076   
1             7.8             0.880         0.00             2.6      0.098   
2             7.8             0.760         0.04             2.3      0.092   
3            11.2             0.280         0.56             1.9      0.075   
4             7.4             0.700         0.00             1.9      0.076   
5             7.4             0.660         0.00             1.8      0.075   
6             7.9             0.600         0.06             1.6      0.069   
7             7.3             0.650         0.00             1.2      0.065   
8             7.8             0.580         0.02             2.0      0.073   
9             7.5             0.500         0.36             6.1      0.071   
10            6.7             0.580         0.08             1.8      0.097   
11            7.5             0.500         0.36             6.1      0.071   
12            5.6             0.615         0.00             1.6      0.089   
13            7.8             0.610         0.29             1.6      0.114   
14            8.9             0.620         0.18             3.8      0.176   
15            8.9             0.620         0.19             3.9      0.170   
16            8.5             0.280         0.56             1.8      0.092   
17            8.1             0.560         0.28             1.7      0.368   
18            7.4             0.590         0.08             4.4      0.086   
19            7.9             0.320         0.51             1.8      0.341   
20            8.9             0.220         0.48             1.8      0.077   
21            7.6             0.390         0.31             2.3      0.082   
22            7.9             0.430         0.21             1.6      0.106   
23            8.5             0.490         0.11             2.3      0.084   
24            6.9             0.400         0.14             2.4      0.085   
25            6.3             0.390         0.16             1.4      0.080   
26            7.6             0.410         0.24             1.8      0.080   
27            7.9             0.430         0.21             1.6      0.106   
28            7.1             0.710         0.00             1.9      0.080   
29            7.8             0.645         0.00             2.0      0.082   
30            6.7             0.675         0.07             2.4      0.089   
31            6.9             0.685         0.00             2.5      0.105   
32            8.3             0.655         0.12             2.3      0.083   
33            6.9             0.605         0.12            10.7      0.073   
34            5.2             0.320         0.25             1.8      0.103   
35            7.8             0.645         0.00             5.5      0.086   
36            7.8             0.600         0.14             2.4      0.086   
37            8.1             0.380         0.28             2.1      0.066   
38            5.7             1.130         0.09             1.5      0.172   
39            7.3             0.450         0.36             5.9      0.074   
40            7.3             0.450         0.36             5.9      0.074   
41            8.8             0.610         0.30             2.8      0.088   
42            7.5             0.490         0.20             2.6      0.332   
43            8.1             0.660         0.22             2.2      0.069   
44            6.8             0.670         0.02             1.8      0.050   
45            4.6             0.520         0.15             2.1      0.054   
46            7.7             0.935         0.43             2.2      0.114   
47            8.7             0.290         0.52             1.6      0.113   
48            6.4             0.400         0.23             1.6      0.066   
49            5.6             0.310         0.37      

from dataframe result above, we can see that all the data values are numeric. This proof is also strengthened by this method below:

In [16]:
wineDataset.value_counts()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  free sulfur dioxide  total sulfur dioxide  density  pH    sulphates  alcohol  quality
7.5            0.510             0.02         1.7             0.084      13.0                 31.0                  0.99538  3.36  0.54       10.5     6          4
6.7            0.460             0.24         1.7             0.077      18.0                 34.0                  0.99480  3.39  0.60       10.6     6          4
7.2            0.360             0.46         2.1             0.074      24.0                 44.0                  0.99534  3.40  0.85       11.0     7          4
               0.695             0.13         2.0             0.076      12.0                 20.0                  0.99546  3.29  0.54       10.1     5          4
8.3            0.650             0.10         2.9             0.089      17.0                 40.0                  0.99803  3.29  0.55       9.5      5          3
                     

From the value counts result above, we can see that the data type is int64. Therefore, we can conclude that there's no string data in the dataset. Also, columns that only contain integer values are "Free Sulfur Dioxide", "Total Sulfur Dioxide", and "Quality". The rest of it all have float values.

In [17]:
wineDataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [18]:
wineDataset.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

In [19]:
wineDataset.shape

(1599, 12)

The wine dataframe shape and data dictionary above show that there are 1599 data samples and 12 variables (11 input/independent (x) and 1 output/dependent (y) variables)

In [20]:
print("\n\n")

# Data Cleaning

In [21]:
wineDataset.isnull().sum() # to check whether there's any missing values in the dataset. If there's any, then we need to do the data cleaning.

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

The result above shows that there's no missing value in the dataset. Henceforth, we don't need to do any data cleaning.

In [22]:
print("\n\n")

# Feature Engineering

Let's take a look on the dataset description below.

In [23]:
wineDataset.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

from the result above:

In [24]:
"""
There are various ranges in the data. If we see the pattern, for example:
a. Fixed Acidity: 4,6 (min) <= 7,1 (25%) <= 7,9 (50%) <= 9,2 (75%) <= 15,9 (max) - dtype: float64
b. Volatile Acidity: 0,12 (min) <= 0,39 (25%) <= 0,52 (50%) <= 0,64 (75%) <= 1,58 (max) - dtype: float64
c. Citric Acid: 0 (min) <= 0,09 (25%) <= 0,26 (50%) <= 0,42 (75%) <= 1 (max) - dtype: float64
d. Residual Sugar: 0,9 (min) <= 1,9 (25%) <= 2,2 (50%) <= 2,6 (75%) <= 15,5 (max) - dtype: float64
e. Chlorides: 0,012 (min) <= 0,07 (25%) <= 0,079 (50%) <= 0,09 (75%) <= 0,611 (max) - dtype: float64
f. Free Sulfur Dioxide: 1 (min) <= 7 (25%) <= 14 (50%) <= 21 (75%) <= 72 (max) - dtype: int64
g. Total Sulfur Dioxide: 6 (min) <= 22 (25%) <= 38 (50%) <= 62 (75%) <= 289 (max) - dtype: int64
h. Density: 0,990070 (min) <= 0,995600 (25%) <= 0,996750 (50%) <= 0,997835 (75%) < 1,004 (max) - dtype: float64
i. pH: 0,12 (min) <= 2,74 (25%) <= 3,21 (50%) <= 3,31 (75%) <= 4,01 (max) - dtype: float64
j. Sulphates: 0,33 (min) <= 0,55 (25%) <= 0,62 (50%) <= 0,73 (75%) <= 2 (max) - dtype: float64
k. Alcohol: 8,4 (min) <= 9,5 (25%) <= 10,2 (50%) <= 11,1 (75%) <= 14,9 (max) - dtype: float64
l. Quality: 3 (min) <= 5 (25%) <= 6 (50% & 75%) <= 8 (max) - dtype: int64
"""

'\nThere are various ranges in the data. If we see the pattern, for example:\na. Fixed Acidity: 4,6 (min) <= 7,1 (25%) <= 7,9 (50%) <= 9,2 (75%) <= 15,9 (max) - dtype: float64\nb. Volatile Acidity: 0,12 (min) <= 0,39 (25%) <= 0,52 (50%) <= 0,64 (75%) <= 1,58 (max) - dtype: float64\nc. Citric Acid: 0 (min) <= 0,09 (25%) <= 0,26 (50%) <= 0,42 (75%) <= 1 (max) - dtype: float64\nd. Residual Sugar: 0,9 (min) <= 1,9 (25%) <= 2,2 (50%) <= 2,6 (75%) <= 15,5 (max) - dtype: float64\ne. Chlorides: 0,012 (min) <= 0,07 (25%) <= 0,079 (50%) <= 0,09 (75%) <= 0,611 (max) - dtype: float64\nf. Free Sulfur Dioxide: 1 (min) <= 7 (25%) <= 14 (50%) <= 21 (75%) <= 72 (max) - dtype: int64\ng. Total Sulfur Dioxide: 6 (min) <= 22 (25%) <= 38 (50%) <= 62 (75%) <= 289 (max) - dtype: int64\nh. Density: 0,990070 (min) <= 0,995600 (25%) <= 0,996750 (50%) <= 0,997835 (75%) < 1,004 (max) - dtype: float64\ni. pH: 0,12 (min) <= 2,74 (25%) <= 3,21 (50%) <= 3,31 (75%) <= 4,01 (max) - dtype: float64\nj. Sulphates: 0,33 (mi

In [25]:
"""
From the dataset description, if we see the pattern, we'll be able to categorize dataset.
The results below is based on the range pattern I see in the data through the dataset description output.
For example, column "Volatile Acidity", "Citric Acid", "Chlorides", "Density", "Sulphates" have the similar data range (overall 0 to 2), henceforth I create a category for those columns.
Also, overall, column "Fixed Acidity", "Residual Sugar", and "Alcohol" have similar value ranges from 0 to 16. Even their maximal values are approximately same, close to 15.
Others have specific range categorization, therefore they can't be combined with other columns.
For instance, Category 6 "Quality" will only need 2 subcategories based on the tips above, while other categories will need more than 2 subcategories.

Category 1: 0 -- 2 (Volatile Acidity, Citric Acid, Chlorides, Density, Sulphates)
Category 2: 2 -- 5 (pH)
Category 3: 0 -- 16 (Fixed Acidity, Residual Sugar, Alcohol)
Category 4: 1 -- 72 (Free Sulfur Dioxide)
Category 5: 5 -- 200 (Total Sulfur Dioxide)
Category 6: 1 -- 10 (Quality)
"""

'\nFrom the dataset description, if we see the pattern, we\'ll be able to categorize dataset.\nThe results below is based on the range pattern I see in the data through the dataset description output.\nFor example, column "Volatile Acidity", "Citric Acid", "Chlorides", "Density", "Sulphates" have the similar data range (overall 0 to 2), henceforth I create a category for those columns.\nAlso, overall, column "Fixed Acidity", "Residual Sugar", and "Alcohol" have similar value ranges from 0 to 16. Even their maximal values are approximately same, close to 15.\nOthers have specific range categorization, therefore they can\'t be combined with other columns.\nFor instance, Category 6 "Quality" will only need 2 subcategories based on the tips above, while other categories will need more than 2 subcategories.\n\nCategory 1: 0 -- 2 (Volatile Acidity, Citric Acid, Chlorides, Density, Sulphates)\nCategory 2: 2 -- 5 (pH)\nCategory 3: 0 -- 16 (Fixed Acidity, Residual Sugar, Alcohol)\nCategory 4: 1

In [26]:
print("\n")

### Category 1: Volatile Acidity, Citric Acid, Chlorides, Density, Sulphates (Overall Range 0 ― 2)

In [27]:
"""
>> I will choose minimum value of min; median of Q1, Q2, and Q3; and maximum value of max for the data categorization
>> I choose median of Q1, Q2, and Q3 since if I choose the maximum value of them, later on one or more column will have only subcategory 0, while only one only has subcategory 0 and 1, which shows an imbalance.
Min
a. Volatile Acidity: 0,12
b. Citric Acid: 0
c. Chlorides: 0,012
d. Density: 0,990070
e. Sulphates: 0,7
>> I pick 0 as the minimum value of min as an initial value for the Category 1 categorization later on

25% (Q1)
a. Volatile Acidity: 0,39
b. Citric Acid: 0,09
c. Chlorides: 0,07
d. Density: 0,995600
e. Sulphates: 0,55
>> I pick 0,39 as the median of Q1

50% (Q2)
a. Volatile Acidity: 0,52
b. Citric Acid: 0,26
c. Chlorides: 0,079
d. Density: 0,996750
e. Sulphates: 0,62
>> I pick 0,52 as the median of Q2

75% (Q3)
a. Volatile Acidity: 0,64
b. Citric Acid: 0,42
c. Chlorides: 0,09
d. Density: 0,997835
e. Sulphates: 0,73
>> I pick 0,64 as the median of Q3

Max
a. Volatile Acidity: 1,58
b. Citric Acid: 1
c. Chlorides: 0,611
d. Density: 1,003690
e. Sulphates: 2
>> I pick 2 as the maximum value of Max
"""

'\n>> I will choose minimum value of min; median of Q1, Q2, and Q3; and maximum value of max for the data categorization\n>> I choose median of Q1, Q2, and Q3 since if I choose the maximum value of them, later on one or more column will have only subcategory 0, while only one only has subcategory 0 and 1, which shows an imbalance.\nMin\na. Volatile Acidity: 0,12\nb. Citric Acid: 0\nc. Chlorides: 0,012\nd. Density: 0,990070\ne. Sulphates: 0,7\n>> I pick 0 as the minimum value of min as an initial value for the Category 1 categorization later on\n\n25% (Q1)\na. Volatile Acidity: 0,39\nb. Citric Acid: 0,09\nc. Chlorides: 0,07\nd. Density: 0,995600\ne. Sulphates: 0,55\n>> I pick 0,39 as the median of Q1\n\n50% (Q2)\na. Volatile Acidity: 0,52\nb. Citric Acid: 0,26\nc. Chlorides: 0,079\nd. Density: 0,996750\ne. Sulphates: 0,62\n>> I pick 0,52 as the median of Q2\n\n75% (Q3)\na. Volatile Acidity: 0,64\nb. Citric Acid: 0,42\nc. Chlorides: 0,09\nd. Density: 0,997835\ne. Sulphates: 0,73\n>> I pi

In [28]:
"""
Henceforth, we'll make subcategories based on minimum value of min; maximum value of Q1, Q2, Q3; and maximum value of Max:
Subcategory 0: 0 ― 0,4
Subcategory 1: (> 0,4) ― 0,6
Subcategory 2: (> 0,6) ― 0,7
Subcategory 3: (> 0,7) ― 2


*Note: I start the subcategory from number 0 since it'll give higher prediction score than starting the subcategory from number 1
"""

"\nHenceforth, we'll make subcategories based on minimum value of min; maximum value of Q1, Q2, Q3; and maximum value of Max:\nSubcategory 0: 0 ― 0,4\nSubcategory 1: (> 0,4) ― 0,6\nSubcategory 2: (> 0,6) ― 0,7\nSubcategory 3: (> 0,7) ― 2\n\n\n*Note: I start the subcategory from number 0 since it'll give higher prediction score than starting the subcategory from number 1\n"

In [29]:
category1 = ["volatile acidity", "citric acid", "chlorides", "density", "sulphates"] # to group the columns into a list that I name as Category 1

In [30]:
# for category1Index in category1: # to categorize Category 1
    # wineDataset.loc[(wineDataset[category1Index] >= 0) & (wineDataset[category1Index] <= 1), category1Index] = 0 # subcategory 0
    # wineDataset.loc[(wineDataset[category1Index] > 1) & (wineDataset[category1Index] <= 2), category1Index] = 1 # subcategory 1

If I code like this:

In [31]:
# for category1Index in category1: # to categorize Category 1
    # wineDataset.loc[(wineDataset[category1Index] >= 0) & (wineDataset[category1Index] <= 0.4), category1Index] = 0 # subcategory 0
    # wineDataset.loc[(wineDataset[category1Index] > 0.4) & (wineDataset[category1Index] <= 0.6), category1Index] = 1 # subcategory 1
    # wineDataset.loc[(wineDataset[category1Index] > 0.6) & (wineDataset[category1Index] <= 0.7), category1Index] = 2 # subcategory 2
    # wineDataset.loc[(wineDataset[category1Index] > 0.7) & (wineDataset[category1Index] <= 2), category1Index] = 3 # subcategory 3

the previous value counts outputs will show me four of five columns will only get Category 0 and 3, while category Density will only get Category 0. Actually for category Density is okay since if you see from the dataset description (min, 25%, 50%, 75%, max), you'll see that the value ranges there are not significant, not even differ equal or more than 1. Now, I'll reanalyze so that each column will have various subcategories and subcategories as small as possible (for example, it's better to have Subcategory 0 and 1 than Subcategory 0 and 3).

In [ ]:
"""
>> I will choose minimum value of min; median of Q1, Q2, and Q3; and maximum value of max for the data categorization
>> I choose median of Q1, Q2, and Q3 since if I choose the maximum value of them, later on one or more column will have only subcategory 0, while only one only has subcategory 0 and 1, which shows an imbalance.
Min
a. Volatile Acidity: 0,12
b. Citric Acid: 0
c. Chlorides: 0,012
d. Density: 0,990070
e. Sulphates: 0,7
>> I pick 0 as the minimum value of min as an initial value for the Category 1 categorization later on

25% (Q1)
a. Volatile Acidity: 0,39
b. Citric Acid: 0,09
c. Chlorides: 0,07
d. Density: 0,995600
e. Sulphates: 0,55
>> I pick 0,39 as the median of Q1

50% (Q2)
a. Volatile Acidity: 0,52
b. Citric Acid: 0,26
c. Chlorides: 0,079
d. Density: 0,996750
e. Sulphates: 0,62
>> I pick 0,52 as the median of Q2

75% (Q3)
a. Volatile Acidity: 0,64
b. Citric Acid: 0,42
c. Chlorides: 0,09
d. Density: 0,997835
e. Sulphates: 0,73
>> I pick 0,64 as the median of Q3

Max
a. Volatile Acidity: 1,58
b. Citric Acid: 1
c. Chlorides: 0,611
d. Density: 1,003690
e. Sulphates: 2
>> I pick 2 as the maximum value of Max
"""

In [ ]:
# for category1Index in category1: # to categorize Category 1
    # wineDataset.loc[(wineDataset[category1Index] >= 0) & (wineDataset[category1Index] <= 0.4), category1Index] = 0 # subcategory 0
    # wineDataset.loc[(wineDataset[category1Index] > 0.6) & (wineDataset[category1Index] <= 0.7), category1Index] = 2 # subcategory 2
    # wineDataset.loc[(wineDataset[category1Index] > 0.7) & (wineDataset[category1Index] <= 2), category1Index] = 3 # subcategory 3

In [32]:
# let's see if the values in Category 1 (Volatile Acidity, Citric Acid, Chlorides, Density, Sulphates) have been categorized
wineDataset.head(10)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4               3.0          0.0             1.9        0.0   
1            7.8               3.0          0.0             2.6        0.0   
2            7.8               3.0          0.0             2.3        0.0   
3           11.2               0.0          3.0             1.9        0.0   
4            7.4               3.0          0.0             1.9        0.0   
5            7.4               3.0          0.0             1.8        0.0   
6            7.9               3.0          0.0             1.6        0.0   
7            7.3               3.0          0.0             1.2        0.0   
8            7.8               3.0          0.0             2.0        0.0   
9            7.5               3.0          0.0             6.1        0.0   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0      3.0  3.51        3.0   
1                 25.0                  67.0      3.0  3.20        3.0   
2                 15.0                  54.0      3.0  3.26        3.0   
3                 17.0                  60.0      3.0  3.16        3.0   
4                 11.0                  34.0      3.0  3.51        3.0   
5                 13.0                  40.0      3.0  3.51        3.0   
6                 15.0                  59.0      3.0  3.30        3.0   
7                 15.0                  21.0      3.0  3.39        3.0   
8                  9.0                  18.0      3.0  3.36        3.0   
9                 17.0                 102.0      3.0  3.35        3.0   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5  
5      9.4        5  
6      9.4        5  
7     10.0        7  
8      9.5        7  
9     10.5        5

In [33]:
wineDataset["volatile acidity"].value_counts()

3.0    1154
0.0     445
Name: volatile acidity, dtype: int64

In [34]:
wineDataset["citric acid"].value_counts()

0.0    1156
3.0     443
Name: citric acid, dtype: int64

In [35]:
wineDataset["chlorides"].value_counts()

0.0    1586
3.0      13
Name: chlorides, dtype: int64

In [36]:
wineDataset["density"].value_counts()

3.0    1599
Name: density, dtype: int64

In [37]:
wineDataset["sulphates"].value_counts()

3.0    1586
0.0      13
Name: sulphates, dtype: int64

In [38]:
# from the six outputs above, 